src/main/java/duke/Duke.java

```diff
@@ -0,0 +1,177 @@
+     */
+    public Duke() {
+        ui = new Ui();
+
+        // load saved tasks from file
+        try {
+            // todo: create data folder first if does not exist
+            storage = new Storage("data/tasks.txt");
+            tasks = new TaskList(storage.loadTasks());
+        } catch (IOException e) {
+            ui.showLoadingError();
+            tasks = new TaskList();
+        }
+    }
+    
+    /**
+     * Constructs a Duke object. An attempt will be made to load in tasks from the file specified. If the file
+     * does not exist or cannot be accessed, an empty task list will be loaded.
+     * 
+     * @param filePath Path of file containing saved tasks.
+     */
+    public Duke(String filePath) {
+        ui = new Ui();
+        
+        // load saved tasks from file
+        try {
+            storage = new Storage(filePath);
+            tasks = new TaskList(storage.loadTasks());
+        } catch (IOException e) {
+            ui.showLoadingError();
+            tasks = new TaskList();
+        }
+    }
+
+    /**
+     * Starts the Duke task management assistant. The user will input commands and the appropriate actions
+     * will be performed. Enter "bye" to exit the program.
+     */
+    public void runCli() {
+        System.out.println(ui.showGreeting());
+
+        label:
+        while (true) {
+            String input = ui.getInput();
+            
+            try {
+                Parser instruction = new Parser(input);
+                Command command = instruction.getCommand();
+
+                switch (command) {
+                case BYE:
+                    break label;
+                case DONE:
+                    setDoneAndShow(tasks, ui, Integer.parseInt(instruction.getParameters()));
+                    break;
+                case TODO:
+                    addTaskAndShow(tasks, ui, new Todo(instruction.getParameters()));
+                    break;
+                case DEADLINE:
+                    addTaskAndShow(tasks, ui,
+                            new Deadline(instruction.getParameters(), instruction.getDate()));
+                    break;
+                case EVENT:
+                    addTaskAndShow(tasks, ui,
+                            new Event(instruction.getParameters(), instruction.getDate()));
+                    break;
+                case LIST:
+                    System.out.println(ui.showTasks(tasks));
+                    break;
+                case DELETE:
+                    deleteTaskAndShow(tasks, ui, Integer.parseInt(instruction.getParameters()));
+                    break;
+                case FIND:
+                    ui.showFound(tasks.findTasks(instruction.getParameters()));
+                default:
+                    ;
+                }
+            } catch (InvalidInstructionException e) {
+                System.out.format("\tError: %s. Please try again.%n%n", e.getMessage());
+            }
+        }
```

Comment:
In this block, you have done a good job in categorizing the commands and executing them correspondingly.

Alternatively, you are suggested to create a `Command` class as well as some children `Command` so you only call a general `execute()` mathod.

Reply:
This comment refers to the block below 🔽